File: c/src/ui/frame.c
#include "frame.h"
#include <stdlib.h>

// Box-drawing characters
static const char *frame_chars[] = {
    "╭", "╮", "╰", "╯", "─", "│", "┬", "┤", "┴", "├"
};
#define UP_LEFT 0
#define UP_RIGHT 1
#define DOWN_LEFT 2
#define DOWN_RIGHT 3
#define HDASH 4
#define VDASH 5
#define FORK_DOWN 6
#define FORK_LEFT 7
#define FORK_UP 8
#define FORK_RIGHT 9

#define FRAME_BORDER_COLOR COLOR_GRAY

Frame *frame_new(u32 w, u32 h, u32 x, u32 y) {
    Frame *f = box(Frame);
    f->w = w;
    f->h = h;
    f->x = x;
    f->y = y;
    f->split[0] = NULL;
    f->split[1] = NULL;
    return f;
}

void frame_free(Frame *f) {
    if (f == NULL) return;

    frame_free(f->split[0]);
    frame_free(f->split[1]);

    for (int i = 0; i < f->widgets.count; i++) {
        if (f->widgets.items[i].free_item) {
            f->widgets.items[i].free_item(f->widgets.items[i].ctx);
        }
    }
    list_free(f->widgets);
    free(f);
}

Frame *(*frame_split(Frame *f, u32 coord, FrameSplit split))[2] {
    if (f == NULL) return NULL;

    if (f->split[0] != NULL || f->split[1] != NULL) {
        // Frame is already split - would panic in original
        return NULL;
    }

    u32 pad = 2;

    switch (split) {
    case FRAME_SPLIT_VERTICAL:
        if (coord >= (f->w - pad) || coord < pad) {
            return NULL;  // Invalid split
        }
        f->split[0] = frame_new(coord, f->h, f->x, f->y);
        f->split[1] = frame_new(f->w - coord + 1, f->h, f->x + coord - 1, f->y);
        break;

    case FRAME_SPLIT_HORIZONTAL:
        if (coord >= (f->h - pad) || coord < pad) {
            return NULL;  // Invalid split
        }
        f->split[0] = frame_new(f->w, coord, f->x, f->y);
        f->split[1] = frame_new(f->w, f->h - coord + 1, f->x, f->y + coord - 1);
        break;

    default:
        return NULL;
    }

    return &f->split;
}

void frame_add_widget(Frame *f, Widget widget) {
    if (f == NULL) return;
    list_append(&f->widgets, widget);
}

void frame_draw(Frame *f, Screen *s, int stage) {
    if (f == NULL || s == NULL) return;

    // Recursively draw child frames
    frame_draw(f->split[0], s, stage);
    frame_draw(f->split[1], s, stage);

    // Stage 1: Draw borders
    if (stage == 1) {
        // Top-left corner
        screen_putc(s, f->x, f->y, frame_chars[UP_LEFT]);
        screen_set_fg_color(s, f->x, f->y, FRAME_BORDER_COLOR);

        // Top edge
        for (u32 i = 1; i < (f->w - 1); i++) {
            screen_putc(s, f->x + i, f->y, frame_chars[HDASH]);
            screen_set_fg_color(s, f->x + i, f->y, FRAME_BORDER_COLOR);
        }

        // Top-right corner
        screen_putc(s, f->x + f->w - 1, f->y, frame_chars[UP_RIGHT]);
        screen_set_fg_color(s, f->x + f->w - 1, f->y, FRAME_BORDER_COLOR);

        // Side edges
        for (u32 i = 1; i < (f->h - 1); i++) {
            screen_putc(s, f->x, f->y + i, frame_chars[VDASH]);
            screen_set_fg_color(s, f->x, f->y + i, FRAME_BORDER_COLOR);
            screen_putc(s, f->x + f->w - 1, f->y + i, frame_chars[VDASH]);
            screen_set_fg_color(s, f->x + f->w - 1, f->y + i, FRAME_BORDER_COLOR);
        }

        // Bottom-left corner
        screen_putc(s, f->x, f->y + f->h - 1, frame_chars[DOWN_LEFT]);
        screen_set_fg_color(s, f->x, f->y + f->h - 1, FRAME_BORDER_COLOR);

        // Bottom edge
        for (u32 i = 1; i < (f->w - 1); i++) {
            screen_putc(s, f->x + i, f->y + f->h - 1, frame_chars[HDASH]);
            screen_set_fg_color(s, f->x + i, f->y + f->h - 1, FRAME_BORDER_COLOR);
        }

        // Bottom-right corner
        screen_putc(s, f->x + f->w - 1, f->y + f->h - 1, frame_chars[DOWN_RIGHT]);
        screen_set_fg_color(s, f->x + f->w - 1, f->y + f->h - 1, FRAME_BORDER_COLOR);
    }
    // Stage 2: Draw split joints
    else if (stage == 2) {
        if (f->split[1]) {
            if (f->y == f->split[1]->y) {
                // Vertical split
                screen_putc(s, f->split[1]->x, f->y, frame_chars[FORK_DOWN]);
                screen_set_fg_color(s, f->split[1]->x, f->y, FRAME_BORDER_COLOR);
                screen_putc(s, f->split[1]->x, f->y + f->h - 1, frame_chars[FORK_UP]);
                screen_set_fg_color(s, f->split[1]->x, f->y + f->h - 1, FRAME_BORDER_COLOR);
            } else {
                // Horizontal split
                screen_putc(s, f->x, f->split[1]->y, frame_chars[FORK_RIGHT]);
                screen_set_fg_color(s, f->x, f->split[1]->y, FRAME_BORDER_COLOR);
                screen_putc(s, f->x + f->w - 1, f->split[1]->y, frame_chars[FORK_LEFT]);
                screen_set_fg_color(s, f->x + f->w - 1, f->split[1]->y, FRAME_BORDER_COLOR);
            }
        }
    }
    // Stage 3: Draw items
    else if (stage == 3) {
        for (int i = 0; i < f->widgets.count; i++) {
            if (f->widgets.items[i].draw) {
                BBox bbox = {
                    .x = f->x + 1,
                    .y = f->y + 1,
                    .w = f->w - 2,
                    .h = f->h - 2,
                };
                f->widgets.items[i].draw(f->widgets.items[i].ctx, s, bbox);
            }
        }
    }
}

void frame_draw_all(Frame *f, Screen *s) {
    frame_draw(f, s, 1);  // borders
    frame_draw(f, s, 2);  // joints
    frame_draw(f, s, 3);  // items
}

File: rust/src/ui/frame.rs
//! Hierarchical frame layout

use crate::tui::{cell::Color, screen::Screen};
use crate::ui::widget::{BBox, Widget};

/// Frame split direction
#[derive(Clone, Copy, Debug)]
pub enum FrameSplit {
    Vertical,
    Horizontal,
}

/// Box-drawing characters
const FRAME_CHARS: [&str; 10] = [
    "\u{256D}", // UP_LEFT ╭
    "\u{256E}", // UP_RIGHT ╮
    "\u{2570}", // DOWN_LEFT ╰
    "\u{256F}", // DOWN_RIGHT ╯
    "\u{2500}", // HDASH ─
    "\u{2502}", // VDASH │
    "\u{252C}", // FORK_DOWN ┬
    "\u{2524}", // FORK_LEFT ┤
    "\u{2534}", // FORK_UP ┴
    "\u{251C}", // FORK_RIGHT ├
];

const UP_LEFT: usize = 0;
const UP_RIGHT: usize = 1;
const DOWN_LEFT: usize = 2;
const DOWN_RIGHT: usize = 3;
const HDASH: usize = 4;
const VDASH: usize = 5;
const FORK_DOWN: usize = 6;
const FORK_LEFT: usize = 7;
const FORK_UP: usize = 8;
const FORK_RIGHT: usize = 9;

const FRAME_BORDER_COLOR: Color = Color::GRAY;

/// Frame node in the layout tree
pub struct Frame {
    x: u32,
    y: u32,
    w: u32,
    h: u32,
    children: Option<(Box<Frame>, Box<Frame>)>,
    widgets: Vec<Box<dyn Widget>>,
}

impl Frame {
    pub fn new(w: u32, h: u32, x: u32, y: u32) -> Self {
        Frame {
            x,
            y,
            w,
            h,
            children: None,
            widgets: Vec::new(),
        }
    }

    #[allow(dead_code)]
    pub fn width(&self) -> u32 {
        self.w
    }
    #[allow(dead_code)]
    pub fn height(&self) -> u32 {
        self.h
    }

    /// Split frame, returns references to the two child frames
    pub fn split(&mut self, coord: u32, direction: FrameSplit) -> (&mut Frame, &mut Frame) {
        let (child1, child2) = match direction {
            FrameSplit::Vertical => {
                let child1 = Frame::new(coord, self.h, self.x, self.y);
                let child2 = Frame::new(self.w - coord + 1, self.h, self.x + coord - 1, self.y);
                (child1, child2)
            }
            FrameSplit::Horizontal => {
                let child1 = Frame::new(self.w, coord, self.x, self.y);
                let child2 = Frame::new(self.w, self.h - coord + 1, self.x, self.y + coord - 1);
                (child1, child2)
            }
        };

        self.children = Some((Box::new(child1), Box::new(child2)));

        let children = self.children.as_mut().unwrap();
        (&mut children.0, &mut children.1)
    }

    pub fn add_widget(&mut self, widget: Box<dyn Widget>) {
        self.widgets.push(widget);
    }

    pub fn draw(&self, screen: &mut Screen) {
        self.draw_stage(screen, 1); // borders
        self.draw_stage(screen, 2); // joints
        self.draw_stage(screen, 3); // widgets
    }

    fn draw_stage(&self, screen: &mut Screen, stage: u8) {
        // Recursively draw children
        if let Some((ref child1, ref child2)) = self.children {
            child1.draw_stage(screen, stage);
            child2.draw_stage(screen, stage);
        }

        match stage {
            1 => self.draw_borders(screen),
            2 => self.draw_joints(screen),
            3 => self.draw_widgets(screen),
            _ => {}
        }
    }

    fn draw_borders(&self, screen: &mut Screen) {
        let (x, y, w, h) = (
            self.x as usize,
            self.y as usize,
            self.w as usize,
            self.h as usize,
        );

        if w == 0 || h == 0 {
            return;
        }

        // Top-left corner
        screen.put_char(x, y, FRAME_CHARS[UP_LEFT]);
        screen.set_fg_color(x, y, FRAME_BORDER_COLOR);

        // Top edge
        for i in 1..(w.saturating_sub(1)) {
            screen.put_char(x + i, y, FRAME_CHARS[HDASH]);
            screen.set_fg_color(x + i, y, FRAME_BORDER_COLOR);
        }

        // Top-right corner
        if w > 1 {
            screen.put_char(x + w - 1, y, FRAME_CHARS[UP_RIGHT]);
            screen.set_fg_color(x + w - 1, y, FRAME_BORDER_COLOR);
        }

        // Side edges
        for i in 1..(h.saturating_sub(1)) {
            screen.put_char(x, y + i, FRAME_CHARS[VDASH]);
            screen.set_fg_color(x, y + i, FRAME_BORDER_COLOR);
            if w > 1 {
                screen.put_char(x + w - 1, y + i, FRAME_CHARS[VDASH]);
                screen.set_fg_color(x + w - 1, y + i, FRAME_BORDER_COLOR);
            }
        }

        // Bottom-left corner
        if h > 1 {
            screen.put_char(x, y + h - 1, FRAME_CHARS[DOWN_LEFT]);
            screen.set_fg_color(x, y + h - 1, FRAME_BORDER_COLOR);
        }

        // Bottom edge
        for i in 1..(w.saturating_sub(1)) {
            if h > 1 {
                screen.put_char(x + i, y + h - 1, FRAME_CHARS[HDASH]);
                screen.set_fg_color(x + i, y + h - 1, FRAME_BORDER_COLOR);
            }
        }

        // Bottom-right corner
        if w > 1 && h > 1 {
            screen.put_char(x + w - 1, y + h - 1, FRAME_CHARS[DOWN_RIGHT]);
            screen.set_fg_color(x + w - 1, y + h - 1, FRAME_BORDER_COLOR);
        }
    }

    fn draw_joints(&self, screen: &mut Screen) {
        if let Some((_, ref child2)) = self.children {
            let (x, y, w, h) = (
                self.x as usize,
                self.y as usize,
                self.w as usize,
                self.h as usize,
            );
            let c2x = child2.x as usize;
            let c2y = child2.y as usize;

            if y == c2y {
                // Vertical split
                screen.put_char(c2x, y, FRAME_CHARS[FORK_DOWN]);
                screen.set_fg_color(c2x, y, FRAME_BORDER_COLOR);
                if h > 1 {
                    screen.put_char(c2x, y + h - 1, FRAME_CHARS[FORK_UP]);
                    screen.set_fg_color(c2x, y + h - 1, FRAME_BORDER_COLOR);
                }
            } else {
                // Horizontal split
                screen.put_char(x, c2y, FRAME_CHARS[FORK_RIGHT]);
                screen.set_fg_color(x, c2y, FRAME_BORDER_COLOR);
                if w > 1 {
                    screen.put_char(x + w - 1, c2y, FRAME_CHARS[FORK_LEFT]);
                    screen.set_fg_color(x + w - 1, c2y, FRAME_BORDER_COLOR);
                }
            }
        }
    }

    fn draw_widgets(&self, screen: &mut Screen) {
        let bbox = BBox {
            x: self.x + 1,
            y: self.y + 1,
            w: self.w.saturating_sub(2),
            h: self.h.saturating_sub(2),
        };

        for widget in &self.widgets {
            widget.draw(screen, bbox);
        }
    }
}
